### PR TITLE
initial commit for code cov workflow updates.

### PR DESCRIPTION
**Pull request by mozrokafor**
initial commit for codecov updates

### CODE REVIEW

File: .github/workflows/macos-build.yaml [hunk 1]
```diff
@@ -117,7 +117,7 @@ jobs:
             CONFIG+=webextension \
             QTPLUGIN+=qsvg \
             CONFIG-=static \
-            CONFIG+=balrog \
+            CONFIG+=balrog \
```
Comment (bakulf) [inline .github/workflows/macos-build.yaml:120]: revert this changes

File: .github/workflows/test_unit.yaml [hunk 2]
```diff
@@ -54,6 +54,28 @@ jobs:
           export PATH=/opt/${{matrix.version}}/gcc_64/bin:$PATH
           ./scripts/tests/unit_tests.sh
 
+      - name: Install Grcov
```
Comment (bakulf) [inline .github/workflows/test_unit.yaml:57]: Ok.. bad news! This approach does not work because the test script "cleans up" the compilation folders. All the gdca files are gone at this point.

We need to work at a script level. I would proceed in this way:

Add a `--grcov file.lcov` param at the script doing something similar to what we did here: https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/scripts/macos/apple_compile.sh#L39-L68

if `--grcov file.lcov` is passed, runs a new function called `utest_grcov` and implement it for `linux/utils/commons.sh`, `macos/utils/commons.sh` and `windows/utils/commons.sh`. For windows, this command will be NOOP. For linux and mac, you run the grcov command to generate the file `file.lcov`.

Then, here, when the script is executed (line 55), pass: `./scripts/tests/unit_test.sh --grcov final.lcov`.

Basically, this complexity must be moved into the script.

**Review by bakulf — CHANGES_REQUESTED** [review]

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
+### Description
+
+    Describe your changes
+
+### Jira Ticket
+
+    i.e https://mozilla-hub.atlassian.net/browse/XXX-####
```
Comment (bakulf) [inline .github/pull_request_template.md:7]: We cannot ask this for community contributions. Say Jira or Github, for community contributions.

File: .github/workflows/test_unit.yaml [hunk 4]
```diff
@@ -44,6 +44,11 @@ jobs:
           export PATH=/opt/6.2.1/gcc_64/bin:$PATH
           python3 scripts/utils/import_languages.py
 
+      - name: Install Grcov
+        shell: bash
+        run: |
+          cargo install grcov
```
Comment (bakulf) [inline .github/workflows/test_unit.yaml:50]: remove this block. We do not use grcov on linux yet.

File: scripts/tests/lottie_tests.sh [hunk 5]
```diff
@@ -10,9 +10,41 @@ if [ -f .env ]; then
   . .env
 fi
 
+helpFunction() {
+  print G "Usage:"
+  print N "\t$0 <macos|ios|macostest> [-d|--debug] [-n|--networkextension] [-a|--adjusttoken <adjust_token>]"
```
Comment (bakulf) [inline scripts/tests/lottie_tests.sh:15]: update this :)

File: scripts/tests/lottie_tests.sh [hunk 6]
```diff
@@ -10,9 +10,41 @@ if [ -f .env ]; then
...
+while [[ $# -gt 0 ]]; do  
+  key="$1"
+  
+  case $key in
+  -g | --grcov)    
+    GRCOV_FILENAME="$2"
+    shift
+    shift
+    ;;
+  -h | --help)
```
Comment (bakulf) [inline scripts/tests/lottie_tests.sh:42]: replace this with `*)`

File: scripts/tests/qml_tests.sh [hunk 7]
```diff
@@ -39,7 +71,20 @@ qmltest_compile || die
...
+  printn Y "Generating temp coverage file for lottie unit tests..."  
+  _grcov_qml qml_test_"$GRCOV_FILENAME"
+fi
+
 printn Y "Cleaning the existing project... "
 qmltest_cleanup || die
 
+if [[ "$GRCOV_FILENAME" ]]; then  
+  printn Y "merging temp files to qml_final_lcov.info... "
+  grcov -t lcov -o qml_final_lcov.info qml_test_"$GRCOV_FILENAME" || die "merging temp files to final failed"
```
Comment (bakulf) [inline scripts/tests/qml_tests.sh:84]: Well... in theory you should use `"$GRCOV_FILENAME"` instead of `qml_final_lcov.info`, right?

File: scripts/tests/unit_tests.sh [hunk 8]
```diff
@@ -10,9 +10,41 @@ if [ -f .env ]; then
   . .env
 fi
 
+helpFunction() {
+  print G "Usage:"
+  print N "\t$0 <macos|ios|macostest> [-d|--debug] [-n|--networkextension] [-a|--adjusttoken <adjust_token>]"
```
Comment (bakulf) [inline scripts/tests/unit_tests.sh:15]: update this!

File: scripts/tests/unit_tests.sh [hunk 9]
```diff
@@ -10,9 +10,41 @@ if [ -f .env ]; then
...
+while [[ $# -gt 0 ]]; do  
+  key="$1"
+  
+  case $key in
+  -g | --grcov)    
+    GRCOV_FILENAME="$2"
+    shift
+    shift
+    ;;
+  -h | --help)
```
Comment (bakulf) [inline scripts/tests/unit_tests.sh:42]: `*)`

File: scripts/tests/unit_tests.sh [hunk 10]
```diff
@@ -45,25 +77,48 @@ utest_compile_unit || die "Failed"
...
+  printn Y "Generating temp coverage file for native messaging unit tests..."  
+  _grcov_unit nativemessaging_"$GRCOV_FILENAME"
+fi
+
+printn Y "Cleaning the existing native messaging project... "
 utest_cleanup_nativemessaging || die "Failed"
 
+if [[ "$GRCOV_FILENAME" ]]; then  
+  printn Y "merging temp files to finallcov.info... "
+  grcov -t lcov -o unit_final_lcov.info unit_"$GRCOV_FILENAME" auth_"$GRCOV_FILENAME" nativemessaging_"$GRCOV_FILENAME" || die "merging temp files to final failed"
```
Comment (bakulf) [inline scripts/tests/unit_tests.sh:118]: `"$GRCOV_FILENAME"` vs `unit_final_lcov.info`

File: scripts/tests/lottie_tests.sh [hunk 11]
```diff
@@ -49,7 +86,20 @@ lottie_compile_qml || die
...
+  printn Y "Generating temp coverage file for lottie qml tests...$GRCOV_FILENAME"  
+  _grcov_lottie_qml lottie_qml_"$GRCOV_FILENAME"
+fi
+
 print Y "Cleaning the existing project... "
 lottie_cleanup_qml || die
 
+if [[ "$GRCOV_FILENAME" ]]; then  
+  printn Y "merging temp files to lottie_final_lcov.info... "
+  grcov -t lcov -o lottie_final_lcov.info lottie_unit_"$GRCOV_FILENAME" lottie_qml_"$GRCOV_FILENAME" || die "merging temp files to final failed"
```
Comment (bakulf) [inline scripts/tests/lottie_tests.sh:99]: `"$GRCOV_FILENAME"` vs `unit_final_lcov.info`

**Review by bakulf — CHANGES_REQUESTED** [review]

File: scripts/tests/unit_tests.sh (hunk 10 above)
Comment (mozrokafor) [reply scripts/tests/unit_tests.sh:118]: i'll rename it to take the given name from the yml file

File: .github/pull_request_template.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
+## Description
+
+    Describe your changes
+
+## Reference
+
+    i.e Jira or Github
```
Comment (bakulf) [inline .github/pull_request_template.md:7]: `Jira or Github issue URL` ?

File: .github/workflows/macos-build.yaml (hunk 1 above)
Comment (bakulf) [reply .github/workflows/macos-build.yaml:114]: This is still here. Remove this and the other one in this file.

File: package.json [hunk 13]
```diff
@@ -1,6 +1,6 @@
 {
   "scripts": {
-    "functionalTest": "mocha --require ./tests/functional/setupVpn.js --timeout 30000 --retries 3"
+    "functionalTest": "mocha --require ./tests/functional/setupVpn.js --reporter json > mocha-results.json  --timeout 30000 --retries 3"
```
Comment (bakulf) [inline package.json:3]: do we use the mocha-result.json? If not, remove this change.

File: scripts/tests/lottie_tests.sh [hunk 14]
```diff
@@ -10,9 +10,33 @@ if [ -f .env ]; then
   . .env
 fi
 
+helpFunction() {
+  print G "Usage:"
+  print N "\t$0 <macos|ios|macostest> [-g|--grcov]"
```
Comment (bakulf) [inline scripts/tests/lottie_tests.sh:15]: update this as discussed on slack.

File: scripts/tests/qml_tests.sh [hunk 15]
```diff
@@ -10,9 +10,33 @@ if [ -f .env ]; then
   . .env
 fi
 
+helpFunction() {
+  print G "Usage:"
+  print N "\t$0 <macos|ios|macostest> [-g|--grcov]"
```
Comment (bakulf) [inline scripts/tests/qml_tests.sh:15]: update this as discussed on slack.

File: scripts/tests/qml_tests.sh [hunk 16]
```diff
@@ -39,7 +63,20 @@ qmltest_compile || die
...
+  printn Y "Generating temp coverage file for lottie unit tests..."  
+  _grcov_qml test_"$GRCOV_FILENAME"
+fi
+
 printn Y "Cleaning the existing project... "
 qmltest_cleanup || die
 
+if [[ "$GRCOV_FILENAME" ]]; then  
+  printn Y "merging temp files to qml_lcov.info... "
+  grcov -t lcov -o "$GRCOV_FILENAME" test_"$GRCOV_FILENAME" || die "merging temp files to final failed"
```
Comment (bakulf) [inline scripts/tests/qml_tests.sh:76]: Well... in this file, there is only 1 GRCOVFILE. Use it directly. Just change line 68 to pass $GRCOV_FILENAME and remove this block.

File: scripts/tests/unit_tests.sh [hunk 17]
```diff
@@ -10,9 +10,33 @@ if [ -f .env ]; then
   . .env
 fi
 
+helpFunction() {
+  print G "Usage:"
+  print N "\t$0 <macos|ios|macostest> [-g|--grcov]"
```
Comment (bakulf) [inline scripts/tests/unit_tests.sh:15]: update this as discussed on slack.

**Review by bakulf — CHANGES_REQUESTED** [review]

File: scripts/tests/qml_tests.sh (hunk 16 above)
Comment (mozrokafor) [reply scripts/tests/qml_tests.sh:76]: good point

File: .github/workflows/macos-build.yaml [hunk 18]
```diff
@@ -110,8 +110,7 @@ jobs:
             CONFIG+=release \
             CONFIG+=webextension \
             QTPLUGIN+=qsvg \
-            CONFIG-=static \
-            CONFIG+=balrog \
+            CONFIG-=static \
```
Comment (bakulf) [inline .github/workflows/macos-build.yaml:113]: mmm what is happening here? are you rebasing the PR?

File: .github/workflows/macos-build.yaml [hunk 19]
```diff
@@ -247,7 +246,6 @@ jobs:
             CONFIG+=webextension \
             QTPLUGIN+=qsvg \
             CONFIG-=static \
-            CONFIG+=balrog \
```
Comment (bakulf) [inline .github/workflows/macos-build.yaml:250]: you should not have changes in this file at all!

File: scripts/tests/qml_tests.sh [hunk 20]
```diff
@@ -39,6 +63,11 @@ qmltest_compile || die
 print Y "Running the QML tests..."
 qmltest_run || die
 
+if [[ "$GRCOV_FILENAME" ]]; then
+  printn Y "Generating temp coverage file for lottie unit tests..."  
+  grcov -t lcov -o "$GRCOV_FILENAME" || die "Failure generating final qml lcov file"
```
Comment (bakulf) [inline scripts/tests/qml_tests.sh:68]: What does this command?
You must generate the grcov file using qmltest_grcov, right?

File: scripts/tests/unit_tests.sh (hunk 9 above)
Comment (bakulf) [reply scripts/tests/unit_tests.sh:34]: yes but it's not done yet :)

File: .github/pull_request_template.md [hunk 21]
```diff
@@ -0,0 +1,24 @@
+## Description
```
Comment (bakulf) [inline .github/pull_request_template.md:1]: This file is a separate thing, right? Maybe you want to land it in a separate PR.

**Review by bakulf — CHANGES_REQUESTED** [review]

File: .github/workflows/test_unit.yaml [hunk 22]
```diff
@@ -43,7 +43,7 @@ jobs:
         run: |
           export PATH=/opt/6.2.1/gcc_64/bin:$PATH
           python3 scripts/utils/import_languages.py
-
+
```
[the remark this answers is not in the document]
Comment (bakulf) [reply .github/workflows/test_unit.yaml:46]: remove this change, please.

File: .github/workflows/macos-build.yaml (hunk 1 above)
Comment (bakulf) [reply .github/workflows/macos-build.yaml:114]: remove these extra spaces.

File: scripts/macos/utils/commons.sh [hunk 23]
```diff
@@ -94,6 +94,10 @@ utest_cleanup_nativemessaging() {
   _cleanup tests.xcodeproj/ ./Release/tests || die
 }
 
+utest_grcov() {    
+  grcov .obj/tests.build/Release/tests.build/Objects-normal/x86_64/  -s . -t lcov --branch --ignore-not-existing . -o "$1" || die "Failure running grcov unit"
```
Comment (bakulf) [inline scripts/macos/utils/commons.sh:98]: I still think having a `_grcov` function is nicer. In here you have a lot of dup code (all these params).
If you implement a simple:
```
_grcov {
  grcov "$2"  -s . -t lcov --branch --ignore-not-existing . -o "$1" || die "Failed to run grcov"
}
```

and then here you run:

`_grcov .obj/tests.build/Release/tests.build/Objects-normal/x86_64/ "$1"`

File: scripts/macos/utils/commons.sh [hunk 24]
```diff
@@ -122,6 +126,14 @@ lottie_cleanup_qml() {
   _cleanup tst_lottie.xcodeproj ./Release/tst_lottie || die
 }
 
+lottie_unit_grcov() {    
+  grcov .obj/lottie_tests.build/Release/lottie_tests.build/Objects-normal/x86_64/  -s . -t lcov --branch --ignore-not-existing . -o "$1" || die "Failure running grcov lottie"
```
Comment (bakulf) [inline scripts/macos/utils/commons.sh:130]: here too

File: scripts/macos/utils/commons.sh [hunk 25]
```diff
@@ -122,6 +126,14 @@ lottie_cleanup_qml() {
...
+}
+
+lottie_qml_grcov() {    
+  grcov .obj/tst_lottie.build/Release/tst_lottie.build/Objects-normal/x86_64/  -s . -t lcov --branch --ignore-not-existing . -o "$1" || die "Failure running grcov qml lottie"
```
Comment (bakulf) [inline scripts/macos/utils/commons.sh:134]: here too

File: scripts/macos/utils/commons.sh [hunk 26]
```diff
@@ -136,3 +148,7 @@ qmltest_run() {
 qmltest_cleanup() {
   _cleanup qml_tests.xcodeproj ./Release/qml_tests || die
 }
+
+qmltest_grcov() {      
+  grcov .obj/qml_tests.build/Release/qml_tests.build/Objects-normal/x86_64/ -s . -t lcov --branch --ignore-not-existing . -o "$1" || die "Failure running grcov qml"
```
Comment (bakulf) [inline scripts/macos/utils/commons.sh:153]: here too

File: scripts/tests/lottie_tests.sh [hunk 27]
```diff
@@ -49,7 +78,20 @@ lottie_compile_qml || die
 print Y "Running the lottie QML tests..."
 lottie_run_qml || die
 
+if [[ "$GRCOV_FILENAME" ]]; then
+  printn Y "Generating temp coverage file for lottie qml tests...$GRCOV_FILENAME"
```
Comment (bakulf) [inline scripts/tests/lottie_tests.sh:82]: why $GRCOV_FILENAME in this string?

File: package.json (hunk 13 above)
Comment (bakulf) [reply package.json:3]: still here.

File: scripts/tests/qml_tests.sh [hunk 28]
```diff
@@ -39,6 +63,11 @@ qmltest_compile || die
 print Y "Running the QML tests..."
 qmltest_run || die
 
+if [[ "$GRCOV_FILENAME" ]]; then
+  printn Y "Generating temp coverage file for qml tests..."  
+  qmltest_grcov "$GRCOV_FILENAME"1wqSAXresfdzvcx ¨.
```
Comment (bakulf) [inline scripts/tests/qml_tests.sh:68]: ?

File: scripts/tests/unit_tests.sh [hunk 29]
```diff
@@ -45,25 +69,48 @@ utest_compile_unit || die "Failed"
...
 print Y "Compile the auth-unit-tests..."
 utest_compile_auth || die "Failed"
 
-print Y "Running the unit-tests..."
+print Y "Running the auth unit-tests..."
 utest_run_auth || die "Failed"
 
-printn Y "Cleaning the existing project... "
+if [[ "$GRCOV_FILENAME" ]]; then
+  printn Y "Generating temp coverage file for auth tests...$GRCOV_FILENAME"
```
Comment (bakulf) [inline scripts/tests/unit_tests.sh:87]: why is the variable used in this string?

**Review by bakulf — CHANGES_REQUESTED** [review]
Not ready yet.

File: scripts/tests/qml_tests.sh (hunk 28 above)
Comment (mozrokafor) [reply scripts/tests/qml_tests.sh:68]: good catch, weird accident

**Review by bakulf — APPROVED** [review]

File: scripts/macos/utils/commons.sh [hunk 30]
```diff
@@ -47,6 +47,10 @@ _compile() {
   [ -f "$2" ] || die "Expected $2 binary"
 }
 
+_grcov() {
+  grcov "$2"  -s . -t lcov --branch --ignore-not-existing . -o "$1" || die "Failed to run grcov"
```
Comment (bakulf) [inline scripts/macos/utils/commons.sh:51]: `grcov "$1"  -s . -t lcov --branch --ignore-not-existing -o "$2"`

**Review by bakulf — APPROVED** [review]
Apply this last change to make it green on try.
